applications/ShapeOptimizationApplication/python_scripts/algorithm_bead_optimization.py

```diff
@@ -0,0 +1,398 @@
+
+                # Filter penalty term if specified
+                if self.filter_penalty_term:
+                    self.mapper.InverseMap(DPDALPHA, DPDALPHA_MAPPED)
+
+                # Compute value of Lagrange function
+                L = objective_value + current_lambda*penalty_value + penalty_factor*(penalty_value)**2
+                if inner_iteration == 1:
+                    dL_relative = 0.0
+                else:
+                    dL_relative = 100*(L/previos_L-1)
+
+                # Compute gradient of Lagrange function
+                if self.filter_penalty_term:
+                    for node in self.design_surface.Nodes:
+                        dLdalpha_i = node.GetSolutionStepValue(DF1DALPHA_MAPPED) + current_lambda*node.GetSolutionStepValue(DPDALPHA_MAPPED)
+                        node.SetSolutionStepValue(DLDALPHA, dLdalpha_i)
+                else:
+                    for node in self.design_surface.Nodes:
+                        dLdalpha_i = node.GetSolutionStepValue(DF1DALPHA_MAPPED) + current_lambda*node.GetSolutionStepValue(DPDALPHA)
+                        node.SetSolutionStepValue(DLDALPHA, dLdalpha_i)
+
+                # Normalization using infinity norm
+                dLdalpha_for_normalization = {}
+                for node in self.design_surface.Nodes:
+                    nodal_alpha = node.GetSolutionStepValue(ALPHA)
+                    if nodal_alpha==self.lower_bound or nodal_alpha==self.upper_bound or node.GetValue(IS_ALPHA_FIXED):
+                        dLdalpha_for_normalization[node.Id] = 0.0
+                    else:
+                        dLdalpha_for_normalization[node.Id] = node.GetSolutionStepValue(DLDALPHA)**2
+
+                max_value = math.sqrt(max(dLdalpha_for_normalization.values()))
+                if max_value == 0:
+                    max_value = 1
+
+                # Compute updated design variable
+                for node in self.design_surface.Nodes:
+                    dalpha = -self.step_size*node.GetSolutionStepValue(DLDALPHA)/max_value
+                    alpha_new = node.GetSolutionStepValue(ALPHA) + dalpha
+
+                    # Enforce bounds
+                    alpha_new = max(alpha_new, self.lower_bound)
+                    alpha_new = min(alpha_new, self.upper_bound)
+
+                    # Enforce constraints
+                    if node.GetValue(IS_ALPHA_FIXED):
+                        alpha_new = 0.0
+
+                    node.SetSolutionStepValue(ALPHA,alpha_new)
+
+                    alpha_new_vectorized = alpha_new * node.GetValue(BEAD_DIRECTION)
+                    node.SetSolutionStepValue(CONTROL_POINT_CHANGE,alpha_new_vectorized)
+
+                # Map design variables
+                self.mapper.Map(ALPHA, ALPHA_MAPPED)
+                for itr in range(self.number_of_filter_iterations-1):
+                    self.mapper.Map(ALPHA_MAPPED, ALPHA_MAPPED)
```

Comment:
Internally we assemble the variables into a vector, so nothing gets overwritten.
This will produce wrong results however for the matrix free mapper. (also above for the inverse mapping)

Reply:
You are right. I will remove this option for multiple filtering, though. Its basically a leftover of some trials.